libqtile/layout/plasma.py

```diff
@@ -0,0 +1,1008 @@
+
+
+class Point(NamedTuple):
+    x: int
+    y: int
+
+
+class Dimensions(NamedTuple):
+    x: int
+    y: int
+    width: int
+    height: int
+
+
+class Orient(Flag):
+    HORIZONTAL = 0
+    VERTICAL = 1
+
+
+class Direction(Enum):
+    UP = auto()
+    DOWN = auto()
+    LEFT = auto()
+    RIGHT = auto()
+
+    @property
+    def orient(self):
+        return Orient.HORIZONTAL if self in [self.LEFT, self.RIGHT] else Orient.VERTICAL
+
+    @property
+    def offset(self):
+        return 1 if self in [self.RIGHT, self.DOWN] else -1
+
+
+class Priority(Enum):
+    FIXED = auto()
+    BALANCED = auto()
+
+
+class AddMode(Flag):
+    HORIZONTAL = 0
+    VERTICAL = 1
+    SPLIT = auto()
+
+    @property
+    def orient(self):
+        return Orient.VERTICAL if self & self.VERTICAL else Orient.HORIZONTAL
+
+
+border_check = {
+    Direction.UP: lambda a, b: isclose(a.y, b.y_end),
+    Direction.DOWN: lambda a, b: isclose(a.y_end, b.y),
+    Direction.LEFT: lambda a, b: isclose(a.x, b.x_end),
+    Direction.RIGHT: lambda a, b: isclose(a.x_end, b.x),
+}
+
+
+def flatten(value):
+    """Flattens a nested list of lists into a single list."""
+    out = []
+    for x in value:
+        if not isinstance(x, list):
+            out.append(x)
+        else:
+            out.extend(flatten(x))
+    return out
+
+
+class Node:
+    """
+    A tree node.
+
+    Each node represents a container that can hold a payload and child nodes.
+    """
+
+    min_size_default = 100
+    root_orient = Orient.HORIZONTAL
+    priority = Priority.FIXED
+
+    def __init__(self, payload=None, x=None, y=None, width=None, height=None):
+        self.payload = payload
+        self._x = x
+        self._y = y
+        self._width = width
+        self._height = height
+        self._size = None
+        self.children = []
+        self.last_accessed = 0
+        self.parent = None
+        self.restorables = {}
+
+    def __repr__(self):
+        info = self.payload or ""
+        if self:
+            info += " +%d" % len(self)
+        return "<Node %s %x>" % (info, id(self))
+
+    def __contains__(self, node):
+        if node is self:
+            return True
+        for child in self:
+            if node in child:
+                return True
+        return False
+
+    def __iter__(self):
+        yield from self.children
+
+    def __getitem__(self, key):
+        return self.children[key]
+
+    def __setitem__(self, key, value):
+        self.children[key] = value
+
+    def __len__(self):
+        return len(self.children)
+
+    @property
+    def root(self):
+        try:
+            return self.parent.root
+        except AttributeError:
+            return self
+
+    @property
+    def is_root(self):
+        return self.parent is None
+
+    @property
+    def is_leaf(self):
+        # self is True if self.children is not an empty list (due to __len__)
+        # self.is_leaf therefore returns True if the node has no children
+        return not self
```

Comment:
This is a little weird; why not just test len(self) instead of having a comment here to explain that we're testing len(self)? :)

Reply:
By having the `__len__` dunder method, the class can be tested for truthiness directly. I just added that note to explain why the existing code worked. But, you're right, we could explicitly check `len(self)`!